ECS Refactor: Alternate Universe

FiniteReality opened the pull request:
Basically #622 but based on a different concept. This is mainly just for comparison.

FiniteReality committed:
Initial move of files

FiniteReality committed:
Switch to using pragma once

FiniteReality committed:
Slowly working through fixing compile errors

FiniteReality committed:
Refactor Configurator to improve modularity

FiniteReality committed:
It now supposedly compiles?

FiniteReality committed:
Things I missed from the CMakeLists

RubyNova committed:
Part one of cmdlist concurrency upgrade. TODO: Update contexts to be created via swapchain image objects.

RubyNova committed:
Fix issue in VulkanGraphicsCmdList where chack was inverted for some reason.

RubyNova committed:
Push for Monica.

FiniteReality committed:
Make context creatable from swapchain image, and render pass creatable from context

RubyNova committed:
Add new ObjectPool type. Add missing memory.hpp header to CMakeLists. Make graphics contexts thread-local in the implemenetation, this still needs to be done on the agnostic side. Delete tons of device code. TODO: Allow Swapchain images to present and to resolve where the queue submit logic will live.

FiniteReality committed:
More cleanup and hooking up the new context doohickery

RubyNova committed:
checkpoint before implementing render targets.

RubyNova committed:
This is as good as its gonna get for now.

FiniteReality committed:
Fix a few linux build failures, fail to fix resize lag

FiniteReality committed:
Allow command lists to be created independently

FiniteReality committed:
Orchestrator attempt 1

FiniteReality committed:
Actually sort the entities beforehand

FiniteReality committed:
Allow creation of independent graphics contexts

FiniteReality committed:
Sample sprite system

FiniteReality committed:
Refactor types a little

FiniteReality committed:
Fix some errors, add a container for efficient(?) render pass mapping

FiniteReality committed:
Make command lists able to be created from the context

RubyNova committed:
Get intellisense working again idk. Add Vulkan cmdlist dispatch support.

RubyNova committed:
Add indexed square to SpriteRendererSystem.

RubyNova committed:
Taking a break. TODO: Add pipeline, make new shaders for sprite rendering, figure out how we're gonna cache descriptor sets.

RubyNova committed:
Port resource loading code. I have no idea if any of it works. It probably should? It worked when I used it last.

RubyNova committed:
Update resource loading code targets.

FiniteReality committed:
Add ECS graphics sample for future testing

FiniteReality committed:
Make it link

RubyNova committed:
Make new sample app be reasonable on Windows under MSVC.

RubyNova committed:
Get the core orchestrator mostly fleshed out. TODO: fix cmdlist caching, then try to make the sample do something.

RubyNova committed:
Code now compiles. VkDevice is getting memstomped. I think it's because i've tried to take a shortcut. Sadface.

RubyNova committed:
The clobbering continues. I think.

FiniteReality committed:
Fix recursive swapchain bug, improve ECS builders, begin sample refactor

RubyNova committed:
un-destroy the sample system. TODO: Now getting FileNotFound as planned.

RubyNova committed:
Add SPV resources.

RubyNova committed:
Fix missing linked list component.

RubyNova committed:
Vulkan API update. GEt secondary cmdlists working.

FiniteReality committed:
Triangle get! Removes some extraneous logging, indentation changes, etc.

FiniteReality committed:
Enable maintenance7/nested command buffers

We probably need a better API for this at some point...

RubyNova committed:
Minor tweaks.

RubyNova committed:
Fix VkCommandBuffer memleak.

RubyNova committed:
Fix memory leak two.

FiniteReality committed:
Skip RemoveComponent, it's causing issues

RubyNova committed:
Fix sync issue.

RubyNova committed:
fix vulkanrender sample.

RubyNova committed:
Try to make macOS CI cooperate.

capnkenny committed:
Enable verbose clang-format

capnkenny committed:
Fix verbose flag

capnkenny committed:
Additional param

FiniteReality committed:
Apply clang-format to all source files

FiniteReality committed:
Ignore some clang-tidy warnings

FiniteReality committed:
Ignore imgui sample entirely for now

capnkenny commented on Audio/OpenAL/OpenALAudioProvider.cpp:
```diff
@@ -86,7 +86,7 @@ namespace NovelRT::Audio
     {
         using CallbackType = void ALC_APIENTRY(void* userptr, char level, const char* message, int length) noexcept;
-        using FnType = void ALC_APIENTRY(CallbackType* callback, void* userptr);
+        using FnType = void ALC_APIENTRY(CallbackType * callback, void* userptr);
```
Nit: inconsistent formatting, `CallbackType*` is desired instead of a space between * and the type

capnkenny commented on Ecs/Core/include/NovelRT/Ecs/LinkedEntityListView.hpp:
```diff
@@ -1,24 +1,32 @@
-            LinkedEntityListNodeComponent _currentComponentNode;
-            Catalogue& _catalogue;
+            Components::LinkedEntityListNodeComponent _currentComponentNode;
+            Catalogue* _catalogue;
```
Any reason on switching from ref to ptr?

FiniteReality replied:
A pointer can be trivially copied, which makes using the iterator a lot simpler. Using a ref here would require consideration for copy/move operations, which long-term is probably a good idea to do but for now doesn't seem to be necessary.

capnkenny commented on Ecs/Core/include/NovelRT/Ecs/SystemSchedulerBuilder.hpp:
```diff
@@ -0,0 +1,72 @@
+                configure(scheduler);
+            }
+
+            scheduler.SpinThreads();
```
Do we mean to spin the threads once when building the scheduler? If so, then why?

FiniteReality replied:
SpinThreads is... poorly named - we should change it to something like "InitializeThreads"

capnkenny commented on Ecs/Core/EntityGraphView.cpp:
```diff
@@ -1,28 +1,39 @@
-                                     EntityGraphComponent component) noexcept
-        : _catalogue(catalogue),
+                                     Components::EntityGraphComponent component) noexcept
+        : _catalogue(&catalogue),
```
We're grabbing a ref from a ref in the ctor here? What's going on?

FiniteReality replied:
That's the address-of operator - we're turning the ref into a ptr.

capnkenny commented on Ecs/Core/EntityGraphView.cpp:
```diff
@@ -46,11 +58,11 @@ namespace NovelRT::Ecs
         }
 
-        for (EntityId child : _childrenChanges.getActual())
+        for (EntityId child : _childrenChanges.Get())
```
Nit: is this supposed to be `Get` or `get`? Unsure where we left off with naming conventions

RubyNova replied:
The former, so `Get`

capnkenny commented on Ecs/Core/SystemSchedulerBuilder.cpp:
```diff
@@ -0,0 +1 @@
+
```
LOL????

capnkenny replied:
If we need this for compilation it's fine - otherwise pls clean up?

FiniteReality replied:
Yeah. Empty source files in header-only components are there because CMake needs them for... Reasons:tm:

This one likely had some code in it at some point, but it was moved and the file was simply never renamed.

capnkenny commented on Audio/OpenAL/OpenALAudioProvider.cpp:
```diff
@@ -86,7 +86,7 @@ namespace NovelRT::Audio
     {
         using CallbackType = void ALC_APIENTRY(void* userptr, char level, const char* message, int length) noexcept;
-        using FnType = void ALC_APIENTRY(CallbackType* callback, void* userptr);
+        using FnType = void ALC_APIENTRY(CallbackType * callback, void* userptr);
```
Nit: calling this out due to inconsistency in formatting. ptr symbol should be next to type, not spaced from - I understand if we cannot fix it due to an issue with the formatter/tidier.

FiniteReality replied:
This seems to be a duplicate comment?

capnkenny commented on Ecs/Core/EntityGraphView.cpp:
```diff
@@ -1,28 +1,39 @@
-                                     EntityGraphComponent component) noexcept
-        : _catalogue(catalogue),
+                                     Components::EntityGraphComponent component) noexcept
+        : _catalogue(&catalogue),
```
Grabbing a ref on a ref? What is going on here?

FiniteReality replied:
This seems to be a duplicate comment?

capnkenny commented on Ecs/Core/SystemSchedulerBuilder.cpp:
```diff
@@ -0,0 +1 @@
+
```
LOL?
Seems empty, do we need this just to make the compiler happy during compile or link time?

FiniteReality replied:
This seems to be a duplicate comment?

capnkenny commented on Ecs/Graphics/include/NovelRT/Ecs/Graphics/Components/Sprite.hpp:
```diff
@@ -0,0 +1,27 @@
+
+        [[nodiscard]] inline bool operator==(const Sprite& /* other */) const noexcept
+        {
+            // TODO: implement this
```
Does this need to be implemented right away?

FiniteReality replied:
No, it was moreso intended as an illustration to @RubyNova that got left behind. We'll be wanting this in the future anyway.

capnkenny commented on Ecs/Graphics/include/NovelRT/Ecs/Graphics/EcsGraphicsBuilder.hpp:
```diff
@@ -0,0 +1,117 @@
+
+        void operator()(SystemScheduler& scheduler)
+        {
+            // TODO: add systems and components for rendering
```
Do we need anything else here on round 1?

FiniteReality replied:
We'll likely not have any built-in systems for rendering to start with, adding those in a follow-up PR to section out the work.

capnkenny commented on Ecs/Graphics/include/NovelRT/Ecs/Graphics/RenderOrchestratorSystem.hpp:
```diff
@@ -0,0 +1,179 @@
+                    static_cast<uint32_t>(surface->GetHeight()));
+
+                NovelRT::Graphics::ClearValue colourDataStruct{};
+                colourDataStruct.colour = NovelRT::Graphics::RGBAColour(0, 0, 255, 255);
```
Should this colour be configurable? Or will blue screen be our default?

RubyNova replied:
This will be part of additional API improvements that will need to get done once this is in but the PR is so huge as it is

capnkenny commented on Ecs/Graphics/include/NovelRT/Ecs/Graphics/RenderOrchestratorSystem.hpp:
```diff
@@ -0,0 +1,179 @@
+            image->QueueSubmit(cmdList);
+
+            // TODO: should this be in ECS?
+            _graphicsDevice->PresentFrame();
```
I'd like an answer too, pls :)

RubyNova replied:
For now, yes, this is a future note to myself when things are more settled

capnkenny commented on Ecs/Graphics/include/NovelRT/Ecs/Graphics/SpriteRendererSystem.hpp:
```diff
@@ -0,0 +1,164 @@
+                auto cmdList = context->CreateCmdList({});
+
+                cmdList->Begin();
+                // rendering la la la
+                cmdList->CmdBindPipeline(nullptr); // TODO: sort this
```
Does this need sorting now?

FiniteReality replied:
No, it was moreso intended as an illustration to @RubyNova that got left behind. We'll be wanting this in the future anyway.

capnkenny commented on Ecs/Graphics/include/NovelRT/Ecs/Graphics/SpriteRendererSystem.hpp:
```diff
@@ -0,0 +1,164 @@
+
+                cmdList->CmdBindVertexBuffers(0, 1, buffers, offsets);
+                cmdList->CmdBindIndexBuffer(_indexRegion, NovelRT::Graphics::IndexType::UInt16);
+                // cmdList->CmdBindDescriptorSets(nullptr); // TODO: sort this
```
How about this?

FiniteReality replied:
This was meant as an illustration to @RubyNova - we'll be implementing this in the future though.

capnkenny commented on Ecs/Graphics/include/NovelRT/Ecs/Graphics/SpriteRendererSystem.hpp:
```diff
@@ -0,0 +1,164 @@
+                cmdList->CmdBindVertexBuffers(0, 1, buffers, offsets);
+                cmdList->CmdBindIndexBuffer(_indexRegion, NovelRT::Graphics::IndexType::UInt16);
+                // cmdList->CmdBindDescriptorSets(nullptr); // TODO: sort this
+                cmdList->CmdDrawIndexed(6, 1, 0, 0, 0);
```
Magic numbers? How would these numbers work once sprite support is added as per the _gigantic_ commented out code block?

RubyNova replied:
It should, in theory, draw a flat square plane mesh that we can slap our cute anime characters on to

capnkenny commented on Graphics/Core/include/NovelRT/Graphics/SecondaryCmdListInfo.hpp:
```diff
@@ -0,0 +1,19 @@
+        std::shared_ptr<GraphicsRenderPass<TBackend>> renderPass = nullptr;
+        size_t subpassIndex = 0;
+    };
+}
```
Nit: EOF line missing

capnkenny commented on Graphics/Vulkan/VulkanGraphicsCmdList.cpp:
```diff
@@ -55,50 +122,53 @@ namespace NovelRT::Graphics
-                        returnStruct.color.float32[1] = colour.getGScalar();
-                        returnStruct.color.float32[2] = colour.getBScalar();
-                        returnStruct.color.float32[3] = colour.getAScalar();
+                           returnStruct.color.float32[0] = colour.getRScalar();
+                           returnStruct.color.float32[1] = colour.getGScalar();
+                           returnStruct.color.float32[2] = colour.getBScalar();
+                           returnStruct.color.float32[3] = colour.getAScalar();
```
Maybe a future thought - is there a way to optimize this a bit? Or do we think/know the compiler will prob. just optimize this anyways?

RubyNova replied:
The only optimisation we could maybe make is double check the maths or make the colour data scalar by default.

capnkenny commented on Graphics/Vulkan/VulkanGraphicsContext.cpp:
```diff
@@ -203,98 +117,28 @@ namespace NovelRT::Graphics
     }
 
-    void VulkanGraphicsContext::OnGraphicsSurfaceSizeChanged(Maths::GeoVector2F /*newSize*/)
+    void VulkanGraphicsContext::EndFrame() // TODO: WTF?
```
👀

RubyNova replied:
Yeah don't mind this one lmao this is a long story, still figuring out what to do with this

capnkenny commented on Graphics/Vulkan/VulkanGraphicsSwapchainImage.cpp:
```diff
@@ -0,0 +1,139 @@
+
+        // TODO: submit to the correct queue if it's just transfers somehow?
+        const VkResult queueSubmitResult = vkQueueSubmit(GetDevice()->GetVulkanGraphicsQueue(), 1, &submitInfo,
+                                                         _queueSubmissionFence->GetVulkanFence());
```
Does this need handling now?

RubyNova replied:
No, if I do it now idk how long it'll take anyway

capnkenny reviewed (commented):
Got some questions and nits, will continue reviewing in a while

FiniteReality committed:
Add nolint regex to clang-format, respond to nits

FiniteReality committed:
Use a different workaround instead

FiniteReality committed:
Revert "Fix sync issue."

This reverts commit 92017de3ce9ccfc4e9be3c0c0adc6b549fcaf285.

FiniteReality committed:
Fix swapchain :sync issues once and for all

FiniteReality committed:
Override the thread-count to require single-threaded operation

This is a workaround until we get the scheduler dependency graph implemented.

RubyNova committed:
Add missing include.

RubyNova committed:
Fix sizing issue that clang missed for some reason. Swap extension names around to see if that improves anything.

capnkenny commented on ThirdParty/CMakeLists.txt:
```diff
@@ -102,8 +102,8 @@ thirdparty_module(stduuid
 thirdparty_module(TBB
-  URL https://github.com/oneapi-src/oneTBB/archive/refs/tags/v2021.5.0.zip
-  URL_HASH SHA512=fde7eb56b46fa86b619eb60e587728d19f31639ec06dfd5d7af0dab00ffba38606d7a89dcf050c4cdfa8896560a7424a3431619108f395348dadf60ad6b67a1b
+  URL https://github.com/uxlfoundation/oneTBB/archive/refs/tags/v2022.1.0.zip
```
I believe this will need a readme update - as well as the Vulkan 1.4 updates

capnkenny reviewed (changes requested):
One nit that needs fixing in the README, as well as CI build failures and clang-format need to be addressed first.

FiniteReality committed:
Fix validation and compile errors in VulkanRender sample

Removed RegisterDescriptorSetForFrame or whatever its name was, temporarily.
Ideally in the future we'd have the GraphicsContext store the semaphore-based caching
logic we have now, but in the interest of getting this PR merged we're just implementing
it directly.

FiniteReality committed:
Apply clang-format to files

capnkenny reviewed (approved):
LGTM - thank you for the hard work!
